### PR TITLE
turn on esent only on a large solution. otherwise, just hold everything ...

### PR DESCRIPTION
**Pull request by heejaechang**
...in memory.

threshold to turn on esent is 50MB > solution source size. esent will be turned on dynamically, once turned on, it will turned off again when solution is closed.

this PR addresses #910

### CODE REVIEW

**Comment by heejaechang** [comment]
@pharring @rchande can you take a look.

File: src/VisualStudio/Core/Def/Implementation/ProjectSystem/VisualStudioWorkspaceImpl.cs [hunk 1]
```diff
@@ -115,6 +120,14 @@ internal VisualStudioProjectTracker ProjectTracker
             }
         }
 
+        internal SolutionSizeTracker SolutionSizeTracker
+        {
+            get
+            {
+                return _solutionSizeTracker;
```
Comment (pharring) [inline src/VisualStudio/Core/Def/Implementation/ProjectSystem/VisualStudioWorkspaceImpl.cs]: You can use expression bodied member.
i.e. ```internal SolutionSizeTracker SolutionSizeTracker => _solutionSizeTracker;```

**Comment by rchande** [comment]
:+1:

**Comment by pharring** [comment]
This might impact solution load or project reload times because it's trying to read document text eagerly. When we discussed this yesterday, I thought you were going to use the Solution Crawler to get the sizes. Is that not possible?

**Comment by heejaechang** [comment]
We can do that as well


Sent from my Windows Phone
________________________________

File: src/VisualStudio/Core/Def/ServicesVisualStudio.csproj [hunk 2]
```diff
@@ -64,6 +64,7 @@
     <Compile Include="Implementation\Watson\WatsonErrorReport.cs" />
```
Comment (VladimirReshetnikov) [inline src/VisualStudio/Core/Def/ServicesVisualStudio.csproj]: Consider adding ` <!-- Copyright (c)  Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information. -->` at the top of the file.

File: src/VisualStudio/Core/Def/Implementation/ProjectSystem/VisualStudioWorkspaceImpl.cs [hunk 3]
```diff
@@ -41,6 +41,9 @@ internal abstract class VisualStudioWorkspaceImpl : VisualStudioWorkspace
         private readonly IVsUIShellOpenDocument _shellOpenDocument;
         private readonly IVsTextManager _textManager;
 
+        // solution size tracker
```
Comment (VladimirReshetnikov) [inline src/VisualStudio/Core/Def/Implementation/ProjectSystem/VisualStudioWorkspaceImpl.cs]: The comment seems redundant.

File: src/VisualStudio/Core/Def/Implementation/ProjectSystem/VisualStudioWorkspaceImpl.cs (hunk 1 above)
Comment (VladimirReshetnikov) [inline src/VisualStudio/Core/Def/Implementation/ProjectSystem/VisualStudioWorkspaceImpl.cs]: ... or a read-only property `internal SolutionSizeTracker SolutionSizeTracker { get; }`

File: src/VisualStudio/Core/Def/Implementation/Workspace/SolutionSizeTracker.cs [hunk 4]
```diff
@@ -0,0 +1,266 @@
...
+                    return;
+            }
+        }
+
+        private void DocumentChanged(Document oldDocument, Document newDocument)
+        {
+            _queue.ScheduleTask(async () =>
+            {
+                var sum = await GetDocumentChangeSizeAsync(oldDocument, newDocument).ConfigureAwait(false);
+                _size = Math.Max(_size + sum, 0);
```
Comment (VladimirReshetnikov) [inline src/VisualStudio/Core/Def/Implementation/Workspace/SolutionSizeTracker.cs]: Can `_size + sum` be negative?

**Comment by pharring** [comment]
What I meant by my comment is that this implementation appears to get the text for every document in every project in the solution. While that might be accurate, it would be a lot of work for large solutions. I liked your idea of passively monitoring documents as they're encountered by the Solution Crawler. There might be some book-keeping (to track which documents have already been included in the total), but the overall impact should be much lower since we're reading those documents anyway.

File: src/VisualStudio/Core/Def/Implementation/ProjectSystem/VisualStudioWorkspaceImpl.cs (hunk 3 above)
Comment (heejaechang) [inline src/VisualStudio/Core/Def/Implementation/ProjectSystem/VisualStudioWorkspaceImpl.cs]: done

File: src/VisualStudio/Core/Def/Implementation/ProjectSystem/VisualStudioWorkspaceImpl.cs (hunk 1 above)
Comment (heejaechang) [inline src/VisualStudio/Core/Def/Implementation/ProjectSystem/VisualStudioWorkspaceImpl.cs]: done

File: src/VisualStudio/Core/Def/Implementation/Workspace/SolutionSizeTracker.cs (hunk 4 above)
Comment (heejaechang) [inline src/VisualStudio/Core/Def/Implementation/Workspace/SolutionSizeTracker.cs]: changed size can be + or -. ex) user deleted text in a file then delta will be -

**Comment by heejaechang** [comment]
@pharring you are right, I shouldn't do that. I changed to use FileInfo.Length since I only care about size not content. I decide not to use solution crawler first. if it really is an issue if FileInfo.Length then I will think how to merge this with solution crawler.

File: src/Features/Core/SolutionCrawler/WorkCoordinator.cs [hunk 5]
```diff
@@ -244,12 +244,16 @@ private void ProcessEvents(WorkspaceChangeEventArgs args, IAsyncToken asyncToken
 
             private void OnDocumentOpened(object sender, DocumentEventArgs e)
             {
-                EnqueueWorkItem(e.Document, InvocationReasons.DocumentOpened);
+                var asyncToken = _listener.BeginAsyncOperation("OnDocumentOpened");
```
Comment (heejaechang) [inline src/Features/Core/SolutionCrawler/WorkCoordinator.cs]: this is a bit unrelated to esent but found while working on this. I missed to put this in event queue so work ends up happened on UI thread where workspace event is happening.

**Comment by heejaechang** [comment]
@pharring @VladimirReshetnikov updated PR. can you take a look?

**Comment by heejaechang** [comment]
@pharring can you take a look?

File: src/VisualStudio/Core/Def/Implementation/SolutionSize/SolutionSizeTracker.cs [hunk 6]
```diff
@@ -0,0 +1,186 @@
...
+namespace Microsoft.VisualStudio.LanguageServices.Implementation.SolutionSize
+{
+    /// <summary>
+    /// Track approximate solution size.
+    /// </summary>
+    [Export]
+    [ExportIncrementalAnalyzerProvider(WorkspaceKind.Host), Shared]
+    internal class SolutionSizeTracker : IIncrementalAnalyzerProvider
+    {
+        private readonly IncrementalAnalyzer tracker = new IncrementalAnalyzer();
```
Comment (pharring) [inline src/VisualStudio/Core/Def/Implementation/SolutionSize/SolutionSizeTracker.cs]: Should be _tracker

File: src/VisualStudio/Core/Def/Implementation/SolutionSize/SolutionSizeTracker.cs [hunk 7]
```diff
@@ -0,0 +1,186 @@
...
+
+        public long GetSolutionSize(Workspace workspace, SolutionId solutionId)
+        {
+            var vsWorkspace = workspace as VisualStudioWorkspaceImpl;
+            if (vsWorkspace == null)
+            {
+                return -1;
+            }
+
+            return tracker.GetSolutionSize(solutionId);
```
Comment (pharring) [inline src/VisualStudio/Core/Def/Implementation/SolutionSize/SolutionSizeTracker.cs]: Could be combined like:
```return workspace is VisualStudioWorkspaceImpl ? _tracker.GetSolutionSize(solutionId) : -1```

File: src/VisualStudio/Core/Def/Implementation/SolutionSize/SolutionSizeTracker.cs [hunk 8]
```diff
@@ -0,0 +1,186 @@
...
+
+        IIncrementalAnalyzer IIncrementalAnalyzerProvider.CreateIncrementalAnalyzer(Workspace workspace)
+        {
+            var vsWorkspace = workspace as VisualStudioWorkspaceImpl;
+            if (vsWorkspace == null)
+            {
+                return null;
+            }
+
+            return tracker;
```
Comment (pharring) [inline src/VisualStudio/Core/Def/Implementation/SolutionSize/SolutionSizeTracker.cs]:
```return workspace is VisualStudioWorkspaceImpl ? _tracker : null;

File: src/VisualStudio/Core/Def/Implementation/SolutionSize/SolutionSizeTracker.cs [hunk 9]
```diff
@@ -0,0 +1,186 @@
...
+        }
+
+        private class IncrementalAnalyzer : IIncrementalAnalyzer
+        {
+            private readonly ConcurrentDictionary<ProjectId, long> _map = new ConcurrentDictionary<ProjectId, long>(concurrencyLevel: 2, capacity: 10);
```
Comment (pharring) [inline src/VisualStudio/Core/Def/Implementation/SolutionSize/SolutionSizeTracker.cs]: Thanks for setting the concurrencyLevel. :-)

File: src/VisualStudio/Core/Def/Implementation/SolutionSize/SolutionSizeTracker.cs [hunk 10]
```diff
@@ -0,0 +1,186 @@
...
+            private long _size;
+
+            public long GetSolutionSize(SolutionId solutionId)
+            {
+                if (_solutionId != solutionId)
+                {
+                    return -1;
+                }
+
+                return _size;
```
Comment (pharring) [inline src/VisualStudio/Core/Def/Implementation/SolutionSize/SolutionSizeTracker.cs]: ```return _solutionId == solutionId ? _size : -1;```

File: src/VisualStudio/Core/Def/Implementation/SolutionSize/SolutionSizeTracker.cs [hunk 11]
```diff
@@ -0,0 +1,186 @@
...
+                return SpecializedTasks.EmptyTask;
+            }
+
+            public async Task AnalyzeProjectAsync(Project project, bool semanticsChanged, CancellationToken cancellationToken)
+            {
+                var sum = await GetProjectSizeAsync(project, cancellationToken).ConfigureAwait(false);
+
+                _map.AddOrUpdate(project.Id, sum, (id, existing) => sum);
+
+                _size = _map.Values.Sum(v => v);
```
Comment (pharring) [inline src/VisualStudio/Core/Def/Implementation/SolutionSize/SolutionSizeTracker.cs]: Do you need the ```v => v``` here? There's an overload of .Sum that takes no arguments and appears to do the right thing for IEnumerable<long>

File: src/VisualStudio/Core/Def/Implementation/Workspace/PersistentStorageService.cs [hunk 12]
```diff
@@ -119,6 +146,38 @@ public IPersistentStorage GetStorage(Solution solution)
...
+                return true;
+            }
+
+            // we only use esent for primary solution. (Ex, forked solution will not use esent)
+            if (solution.BranchId != solution.Workspace.PrimaryBranchId || solution.FilePath == null)
+            {
+                return false;
+            }
+
+            return true;
```
Comment (pharring) [inline src/VisualStudio/Core/Def/Implementation/Workspace/PersistentStorageService.cs]: ```return _testing || (solution.BranchId == solution.Workspace.PrimaryBranchId && solution.FilePath != null);```

File: src/VisualStudio/Core/Def/Implementation/Workspace/PersistentStorageService.cs [hunk 13]
```diff
@@ -119,6 +146,38 @@ public IPersistentStorage GetStorage(Solution solution)
...
+                return true;
+            }
+
+            if (_solutionSizeTracker == null)
+            {
+                return false;
+            }
+
+            var size = _solutionSizeTracker.GetSolutionSize(solution.Workspace, solution.Id);
+            return size > SolutionSizeThreshold;
```
Comment (pharring) [inline src/VisualStudio/Core/Def/Implementation/Workspace/PersistentStorageService.cs]: ```return _testing || _solutionSizeTracker?.GetSolutionSize(solution.Workspace, solution.Id) > SolutionSizeThreshold;```

(There is an active issue on the compiler to ensure that the last part of this creates optimal code.)

**Comment by pharring** [comment]
:+1:

File: src/VisualStudio/Core/Def/Implementation/SolutionSize/SolutionSizeTracker.cs (hunk 7 above)
Comment (VladimirReshetnikov) [inline src/VisualStudio/Core/Def/Implementation/SolutionSize/SolutionSizeTracker.cs]: Even if you do not want to convert `if` into `? :`, the usage of `as` seems unnecessary. Just use `is`.

File: src/VisualStudio/Core/Def/Implementation/SolutionSize/SolutionSizeTracker.cs [hunk 14]
```diff
@@ -0,0 +1,186 @@
...
+                try
+                {
+                    // just to reduce exception thrown
+                    if (!File.Exists(filepath))
+                    {
+                        return -1;
+                    }
+
+                    var info = new FileInfo(filepath);
+                    return info.Length;
```
Comment (VladimirReshetnikov) [inline src/VisualStudio/Core/Def/Implementation/SolutionSize/SolutionSizeTracker.cs]: Consider rewriting as `return new FileInfo(filepath).Length;`.

File: src/VisualStudio/Core/Def/Implementation/Workspace/Esent/EsentPersistentStorage.cs [hunk 15]
```diff
@@ -47,6 +46,13 @@ internal partial class EsentPersistentStorage : AbstractPersistentStorage
             _esentStorage = new EsentStorage(databaseFile, enablePerformanceMonitor);
         }
 
+        public static string GetDatabaseFile(string workingFolderPath)
+        {
+            Contract.ThrowIfTrue(string.IsNullOrWhiteSpace(workingFolderPath));
```
Comment (VladimirReshetnikov) [inline src/VisualStudio/Core/Def/Implementation/Workspace/Esent/EsentPersistentStorage.cs]: Windows supports folder names consisting only of whitespace characters.

File: src/VisualStudio/Core/Def/Implementation/Workspace/PersistenceServiceFactory.cs [hunk 16]
```diff
@@ -4,20 +4,29 @@
...
+            _solutionSizeTracker = solutionSizeTracker;
+        }
+
         public IWorkspaceService CreateService(HostWorkspaceServices workspaceServices)
         {
             if (_singleton == null)
             {
                 var optionService = workspaceServices.GetService<IOptionService>();
-                System.Threading.Interlocked.CompareExchange(ref _singleton, new PersistentStorageService(optionService), null);
+                System.Threading.Interlocked.CompareExchange(ref _singleton, new PersistentStorageService(optionService, _solutionSizeTracker), null);
```
Comment (VladimirReshetnikov) [inline src/VisualStudio/Core/Def/Implementation/Workspace/PersistenceServiceFactory.cs]: Consider `using System.Threading;`

File: src/VisualStudio/Core/Def/ServicesVisualStudio.csproj [hunk 17]
```diff
@@ -42,6 +42,7 @@
     <Compile Include="Implementation\Options\ExportOptionSerializerAttribute.cs" />
```
Comment (VladimirReshetnikov) [inline src/VisualStudio/Core/Def/ServicesVisualStudio.csproj]: Consider adding `<!-- Copyright (c) Microsoft. All Rights Reserved. Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information. -->` at the top of the file.

File: src/VisualStudio/Core/Def/Implementation/SolutionSize/SolutionSizeTracker.cs (hunk 6 above)
Comment (heejaechang) [inline src/VisualStudio/Core/Def/Implementation/SolutionSize/SolutionSizeTracker.cs]: done

File: src/VisualStudio/Core/Def/Implementation/SolutionSize/SolutionSizeTracker.cs (hunk 7 above)
Comment (heejaechang) [inline src/VisualStudio/Core/Def/Implementation/SolutionSize/SolutionSizeTracker.cs]: done

File: src/VisualStudio/Core/Def/Implementation/SolutionSize/SolutionSizeTracker.cs (hunk 8 above)
Comment (heejaechang) [inline src/VisualStudio/Core/Def/Implementation/SolutionSize/SolutionSizeTracker.cs]: done

File: src/VisualStudio/Core/Def/Implementation/SolutionSize/SolutionSizeTracker.cs (hunk 10 above)
Comment (heejaechang) [inline src/VisualStudio/Core/Def/Implementation/SolutionSize/SolutionSizeTracker.cs]: done

File: src/VisualStudio/Core/Def/Implementation/SolutionSize/SolutionSizeTracker.cs (hunk 11 above)
Comment (heejaechang) [inline src/VisualStudio/Core/Def/Implementation/SolutionSize/SolutionSizeTracker.cs]: :)

File: src/VisualStudio/Core/Def/Implementation/Workspace/PersistentStorageService.cs (hunk 12 above)
Comment (heejaechang) [inline src/VisualStudio/Core/Def/Implementation/Workspace/PersistentStorageService.cs]: this one, I don't want to shorten it. so I will leave it as it is.

File: src/VisualStudio/Core/Def/Implementation/Workspace/PersistentStorageService.cs (hunk 13 above)
Comment (heejaechang) [inline src/VisualStudio/Core/Def/Implementation/Workspace/PersistentStorageService.cs]: this one also I don't want to shorten it.

File: src/VisualStudio/Core/Def/Implementation/SolutionSize/SolutionSizeTracker.cs (hunk 14 above)
Comment (heejaechang) [inline src/VisualStudio/Core/Def/Implementation/SolutionSize/SolutionSizeTracker.cs]: done

File: src/VisualStudio/Core/Def/Implementation/Workspace/Esent/EsentPersistentStorage.cs (hunk 15 above)
Comment (heejaechang) [inline src/VisualStudio/Core/Def/Implementation/Workspace/Esent/EsentPersistentStorage.cs]: working folder is folder from VS not from user.

File: src/VisualStudio/Core/Def/Implementation/Workspace/PersistenceServiceFactory.cs (hunk 16 above)
Comment (heejaechang) [inline src/VisualStudio/Core/Def/Implementation/Workspace/PersistenceServiceFactory.cs]: done

**Comment by heejaechang** [comment]
@dotnet-bot test this please

**Pull request merged**